Add routines for updating QR decompositions

ewmoore opened the pull request:
This pull request goes on top of #4021, accordingly only the last few commits are relevant here.

Here three new functions are added to perform various modifications of qr decompositions: `qr_update`, `qr_insert` and `qr_delete` for performing rank-k updates, adding row(s) or column(s) and removing row(s) or column(s).  

The methods used are essentially those in [1], with some input from [2].

I've endeavored to write this such that updates can be performed in-place as much as possible.  This is, for instance, why the rank-1 update and the rank-k update algorithms are slightly different.  This allows the rank-1 update to support updating `q` and `r` when they are stored in C order.  (In this spirit, it might be worth changing `linalg.qr` to return fortran ordered 'r', since at present it always returns a C ordered `r`,  and a F ordered `q`.  Numpy's qr returns both as C order.)

Edit: both are done (3/18/15)
Some outstanding things:

1.  I suppose I'll need to add a 'check_finite'  flag to these.  
2.  `overwrite_qr` should probably default to `False` instead of true.

References

1.  Golub, G. H. & Loan, C. F. van V. Matrix Computations, 3rd Ed. (Johns Hopkins University Press, 1996).
2.  Hammarling, S. & Lucas, C. Updating the QR factorization and the least squares problem. 1-73 (The University of Manchester, 2008). at <http://eprints.ma.man.ac.uk/1192/>

argriffing commented:
Is this PR still waiting for the changes suggested in https://github.com/scipy/scipy/pull/4021#issuecomment-66709921?

By the way, I noticed that the code includes a workaround for a numpy 1.5.1 limitation.  Now that support for numpy versions < 1.6 have been dropped, this workaround could be removed.

ewmoore commented:
Mostly. It still needs a little work on updating economic decomposition for
column additions. But since as it is now, it depends on #4021, it can't be
merged until that is finished.  It does not have to work that way. Lapack
is always available when building scipy so this could just link directly if
desired.

Some code review would also be good. Even though there haven't been any
comments, I'll bet there are at least some style issues if not actual code
issues.

On Wednesday, January 7, 2015, argriffing <notifications@github.com> wrote:

> Is this PR still waiting for the changes suggested in #4021 (comment)
> <https://github.com/scipy/scipy/pull/4021#issuecomment-66709921>?
>
> By the way, I noticed that the code includes a workaround for a numpy
> 1.5.1 limitation. Now that support for numpy versions < 1.6 have been
> dropped, this workaround could be removed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scipy/scipy/pull/4249#issuecomment-69046960>.
>

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    for 1D arrays these are equivalent.  This is Numpy's gh-2287, fixed in
+    9b8ff38.
+
+    FIXME: Is it worth only applying this for numpy 1.5.1? 
```
That commit is in numpy 1.6.2, which is now the lowest supported version. So this can go.

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+
+cdef inline blas_t nrm2(int n, blas_t* x, int incx) nogil:
+    # since dnrm2 etc are functions, we can't at present call them...
+    # this is mostly cribbed from netlib's dnrm2
```
A little more explanation would help me here. We have f2py wrappers for ``nrm2``, I also see Cython wrappers in gh-4021. Why can't those be used?

ewmoore replied:
It's only really possible to call a fortran subroutine from C.  dnrm2 is a function and at the time #4021 hadn't wrapped any of the functions.  This will become something like the rest of the fused type wrappers now that that dnrm2 is wrapped.

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        scal(m, 0, u, 1)
+        axpy(n, 1, s, 1, s+n, 1)
+        s[n] = 0
+        return False
```
style nitpick: a blank line here for readability

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    s[n] = wpnorm
+    return True
+
+cdef int thin_qr_row_delete(int m, int n, blas_t* q, int* qs, bint qisF, blas_t* r, int* rs, int k, int p_eco, int p_full) nogil:
```
style: wrap long line to <80 chars

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+
+        memset(s+2*n, 0, n*sizeof(blas_t))
+
+        # what happens here...
```
is this a TODO?

ewmoore replied:
It is mostly a point where I don't really know what the right answer is.  When updating an economic mode decomposition, a vector into the null space of Q (equivalently A, if A = QR) is needed.  The papers I've been working from either do, or suggest doing what I do here, but its probably possible to construct an example where this will fail.  It probably needs to have some kind of "try even harder" fallback.  Alternatively, it might never happen that this fails, I'm not really sure yet.

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        swap(m, row(q, qs, j), qs[1], row(q, qs, j-1), qs[1])
+
+cdef int thin_qr_block_row_insert(int m, int n, blas_t* q, int* qs, blas_t* r,
+        int* rs, blas_t* u, int* us, int k, int p) nogil:
```
style: align indentation with opening bracket

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    return 0
+
+cdef void thin_qr_col_insert(int m, int n, blas_t* q, int* qs, blas_t* r,
+        int* rs, blas_t* u, int* us, int k, int p_eco, int p_full,
```
indentation to opening bracket

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+
+        # apply the Q from this small qr to the last (m-(n-p)) columns of q.
+        info = ormqr(side, trans, m, m-(n-p), p, index2(r, rs, n-p, k), rs[1],
+                tau, index2(q, qs, 0, n-p), qs[1], work, lwork)
```
indentation -> opening bracket

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+
+cdef void hessenberg_qr(int m, int n, blas_t* q, int* qs, blas_t* r, int* rs,
+                        int k) nogil:
+    """Reduce an upper hessenberg matrix r, to upper triangluar, starting in
```
typo triangluar

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    cdef cnp.npy_intp size
+    cdef bint qisF
+
+    if q.ndim != 2: raise ValueError('q must be 2d')
```
style: ``raise`` statement to next line

rgommers replied:
and 2x more the same a few lines below

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+
+    return Q, R, typecode, Q.shape[0], R.shape[1], economic
+ 
+cdef void* extract(cnp.ndarray a, int* as):
```
don't call a parameter ``as``, that's a builtin

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    up = extract(u, us)
+
+    size = 2*n
+    s = cnp.PyArray_ZEROS(1, &size, typecode, 1)
```
this allocates memory and can fail

rgommers replied:
same comment for other usages in this file

ewmoore replied:
It returns a python object, I think cython handles checking the output for me.  I'll look at the generate code to double check, but I think this okay.

rgommers replied:
You're right:

      __pyx_t_2 = PyArray_ZEROS(1, (&__pyx_v_size), __pyx_v_typecode, 1);
      if (unlikely(!__pyx_t_2)) {
          __pyx_filename = __pyx_f[0];
          __pyx_lineno = 1163;
          __pyx_clineno = __LINE__;
          goto __pyx_L1_error;
      }
      __Pyx_GOTREF(__pyx_t_2);
      if (!(likely(((__pyx_t_2) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_2, __pyx_ptype_5numpy_ndarray))))) {
          __pyx_filename = __pyx_f[0];
          __pyx_lineno = 1163;
          __pyx_clineno = __LINE__;
          goto __pyx_L1_error;
      }
      __pyx_v_s = ((PyArrayObject *)__pyx_t_2);
      __pyx_t_2 = 0;

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        return u, s[:n+1], rc_d, info
+    if typecode == cnp.NPY_CFLOAT:
+        rc_f = rcond
+        info = reorth(m, n, <float_complex*>qp, qs, qisF, <float_complex*>up, us, <float_complex*>sp, <float_complex*>&rc_f)
```
break long line

rgommers replied:
and same 5 lines lower

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        s[n] = unorm*wnorm
+        return 0
+
+    # if the above check failed, try one reorthogonalization
```
I read this the wrong way around "check failed" in >50% of the time means "enter the if-block and do something to fix it". it's a minor thing, but could be clarified a bit. Or in the block above add a comment that the ``return 0`` is the main path, and the rest happens in case the check (which does ...) failed.

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        axpy(n, 1, s, 1, s+n, 1)
+        scal(n, unorm, s, 1)
+        s[n] = 0
+        return 1
```
blank line here for readability

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        + Q is contiguous.  This function is preferable over simply calling
+        np.dot for two reasons: 1) this output is always in F order, 2) no
+        copies need be made if Q is complex.  Point 2 in particular makes this
+        a good bit faster for complex inputs.
```
if you still remember the timings, might be good to put an approximate factor here. "a good bit" could be 20% or a factor of 5

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+                uvoid = extract(u, us)
+                ldu = us[0]
+            else:
+                u = PyArray_FromArraySafe(u, NULL, cnp.NPY_F_CONTIGUOUS)
```
makes a copy so check return not NULL?

rgommers replied:
comment valid for further ``PyArray_FromArray`` usages as well

rgommers replied:
so this is also OK, same as ``PyArray_ZEROS``, will delete this comment

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+                        col(<double_complex*>qTuvoid, qTus, k), qTus[1])
+                blas_t_2d_conj(m, p, col(<double_complex*>qTuvoid, qTus, k), qTus)
+        else:
+            raise ValueError('1 <= u.ndim <= 2')
```
Maybe expand this a bit for clarity: ``Shape of u is incorrect, should be 1 <= u.ndim <= 2``

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+                (a.strides[j] / a.descr.itemsize) >= limits.INT_MAX:
+            copy = True
+        if a.shape[j] >= limits.INT_MAX:
+            raise ValueError('Input array to large for use with BLAS')
```
typo: to --> too

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    R = PyArray_CheckFromAny(r0, NULL, 0, 0, r_order, NULL)
+
+    if Q.ndim != 2 or R.ndim != 2:
+        raise ValueError('Q and R must be 2d')
```
nitpick: ``2d`` -->  ``2-D``

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    typecode = cnp.PyArray_TYPE(Q)
+
+    if typecode != cnp.PyArray_TYPE(R):
+        raise ValueError('q and r must have the same type')
```
``q and r`` --> ``Q and R``

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+
+    if not (typecode == cnp.NPY_FLOAT or typecode == cnp.NPY_DOUBLE 
+            or typecode == cnp.NPY_CFLOAT or typecode == cnp.NPY_CDOUBLE):
+        raise ValueError('only floatingcomplex arrays supported')
```
``floatingcomplex`` isn't super clear, maybe ``Only arrays with dtypes int32, int64, float32 and float64 are supported``

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+            or typecode == cnp.NPY_CFLOAT or typecode == cnp.NPY_CDOUBLE):
+        raise ValueError('only floatingcomplex arrays supported')
+
+    # we support MxM MxN and MxN NxN
```
would be good to say this in both error messages below: ``"expected (M,M) (M,N) or (M,N) (N,N) but found (...)"``

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        as[1] = 0
+    return cnp.PyArray_DATA(a)
+
+@cython.embedsignature(True)
```
cool, I wasn't aware this existed. we can use this in more places where we now maintain signatures manually

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+def qr_delete(Q, R, k, p=1, which='row', overwrite_qr=True, check_finite=True):
+    """QR downdate on row or column deletions
+
+    If ``A = Q R`` is the qr factorization of A, return the qr factorization
```
would be good to consistently capitalize qr

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    """QR downdate on row or column deletions
+
+    If ``A = Q R`` is the qr factorization of A, return the qr factorization
+    of `A` where `p` rows or columns have been removed starting at row or
```
double backticks around ``A`` here and in the line above

rgommers replied:
same comment for other public docstrings

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    Q : (M, M) or (M, N) array_like
+        Unitary/orthogonal matrix from QR decomposition.
+    R : (M, N) or (N, N) array_like
+        Upper trianglar matrix from QR decomposition.
```
typo trianglar

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    R : (M, N) or (N, N) array_like
+        Upper trianglar matrix from QR decomposition.
+    k : int
+        index of the first row or column to delete.
```
nitpick: capitalize first letter (and same 2 lines lower)

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    overwrite_qr : bool, optional
+        If True, consume Q and R, overwriting their contents with their
+        downdated versions, and returning approriately sized views.  
+        Defaults to True.
```
``qr`` and ``qr_multiply`` default to False. Both for consistency and safety that should probably be the default here as well.

rgommers replied:
and maybe add ``(may improve performance)`` (copied from ``qr``)

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    check_finite : bool, optional
+        Whether to check that the input matrix contains only finite numbers.
+        Disabling may give a performance gain, but may result in problems
+        (crashes, non-termination) if the inputs do contain infinities or NaNs.
```
add ``Default is True.``

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+        Updated unitary/orthogonal factor
+    R1 : ndarray
+        Updated upper triangular factor
+
```
add

    See Also
    --------
    qr, qr_multiply, qr_insert, qr_update

rgommers replied:
and same in ``qr_insert`` and ``qr_update`` docstrings

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+    positive.
+
+    .. versionadded:: 0.16.0
+
```
Adding a ``References`` section with the right subset of the refs in the comments at the top of the file would be good.

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+from libc.string cimport memset
+cimport numpy as cnp
+
+cdef int NPY_ANYORDER = 0  
```
why define this to zero instead of using ``cnp.NPY_ANYORDER`` in the rest of this file?

ewmoore replied:
Is there a `cnp.NPY_ANYORDER`? If there is, yes absolutely.

rgommers replied:
http://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.NPY_ANYORDER 

and it is part of ``Cython/Includes/numpy/__init__.pxd`` in current Cython master (so I assume also in 0.21/22, didn't check)

ewmoore replied:
Ah okay, seems I missed that.

ewmoore replied:
Ah now I remember.  This needs renamed.  NPY_ANYORDER is part of the NPY_ORDER enum, but this flag goes with NPY_ARRAY_F_CONTIGUOUS etc.  The point is to remove some magic zeros from the code below.

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+                    <double_complex*>extract(r1, rs), rs, k1, p1)
+            return q1[p1:, p1:], r1[p1:, :]
+    elif which == 'col':
+        if p1 > 1:
```
this deserves a comment. so only for deleting a column (not a row) there is a special case when deleting a single column?

ewmoore replied:
Yes, that correct.  There isn't a way to remove the rows in one pass, as there is for the columns, so the block row code really just loops over the rows to be removed.  I can add a note.

rgommers commented on scipy/linalg/_decomp_update.pyx:
```diff
@@ -0,0 +1,2416 @@
+            elif typecode == cnp.NPY_DOUBLE:
+                qr_col_delete(m, o, n, <double*>extract(q1, qs), qs, 
+                    <double*>extract(r1, rs), rs, k1)
+            el
```
typo ``coluns``

rgommers replied:
matrix --> matrices

rgommers replied:
there are quite a few of these blocks; the 4 cases for different types are identical as far as I can see. maybe it's a good idea to also use fused types here?

rgommers replied:
nitpick: space after ``(m,``

rgommers replied:
wrap long lines in the ~10 lines above

rgommers replied:
needs a space at the end, or this will read ``wasinvalid``

rgommers replied:
and same 2 lines lower

rgommers replied:
``malloc`` may not mean anything to a Python user that doesn't know C

rgommers replied:
``U`` --> ``u``

rgommers replied:
insert blank line here

rgommers replied:
insert blank line here

rgommers replied:
these are both deprecated API, rename ``NPY`` --> ``NPY_ARRAY``

rgommers replied:
nitpick: ``2d`` --> ``2-D``

rgommers replied:
``len`` is a builtin name, rename

rgommers replied:
this comment is about the block above right? comment needs to be above code

rgommers replied:
space at end of string here and 3 lines lower

rgommers replied:
no ``malloc`` in user-visible message

ewmoore replied:
I would be nice, but I have to specify the type somewhere.  Cython can't dispatch a fused type based on the type of a numpy array or memory view unless it is fixed at compile time, so I don't think I can.  I could replace the duplicated code with some template stuff and generate it before cython gets called if that is deemed a better approach.

rgommers replied:
The approach used in ``_csparsetools.pyx.in`` may be usable here.

ewmoore replied:
This comment is about the `v1 = v1.T` below.  Given that it returns a view it's fine. I'll remove the comment.

ewmoore replied:
Cython's numpy support doesn't have these names.  I could add them to the top of this file and use them if you'd like, but it will likely still warn when compiling.

rgommers replied:
Hmm, never mind then. I noticed before that there's quite a bit to do to update the Numpy support in Cython. That's for another time....

rgommers commented on scipy/linalg/tests/test_decomp_update.py:
```diff
@@ -0,0 +1,1323 @@
+import itertools
```
nitpick: space between stdlib and other imports

ewmoore replied:
Done.

rgommers commented on scipy/linalg/tests/test_decomp_update.py:
```diff
@@ -0,0 +1,1323 @@
+            a = a + 1j * b
+        a = a.astype(self.dtype)
+        q, r = linalg.qr(a, mode=mode)
+        # numpy 1.5.1 np.triu can modify array dtype. So qr of 'f' arrays
```
update comment to 1.6.2 (and is it still true there?)

ewmoore replied:
No, this can go away.

rgommers commented on scipy/linalg/tests/test_decomp_update.py:
```diff
@@ -0,0 +1,1323 @@
+
+        assert_(p > 0)
+
+        # super call set the seed...
```
It's not ``BaseQRdeltas.__init__`` but ``generate()`` that sets the seed, so a tweak to the comment to make that clear would be good.

rgommers commented on scipy/linalg/tests/test_decomp_update.py:
```diff
@@ -0,0 +1,1323 @@
+        assert_raises(ValueError, qr_insert, q[:-2], r, u, 0, 'col')
+        assert_raises(ValueError, qr_insert, q, r, u[1:], 0, 'col')
+
+    def test_integer_input(self):
```
maybe test all dtypes that aren't supported (float16, bool, int8/16/32/64)

rgommers commented on scipy/linalg/tests/test_decomp_update.py:
```diff
@@ -0,0 +1,1323 @@
+    expected = np.dot(q.T.conj(), u)
+    res = _decomp_update._form_qTu(q, u)
+    assert_allclose(res, expected, rtol=rtol, atol=atol)
+ 
```
add:

    if __name__ == "__main__":
         run_module_suite()

rgommers commented on scipy/linalg/setup.py:
```diff
@@ -119,6 +119,31 @@ def configuration(parent_package='',top_path=None):
+                         include_dirs=['.'])
+
+    config.add_extension('_decomp_update',
+                         sources=['_decomp_update.c'])
```
The lines to add in ``linalg/bento.info`` are:

    Extension: _decomp_update
        Sources:
            _decomp_update.c

ewmoore replied:
Done.

rgommers commented:
In the summary of this PR two "outstanding things" are listed, the first is done and can be removed.

rgommers commented:
@ewmoore this PR looks very good!

I've made a full pass through the code for code/style/doc/test issues. I didn't actually review the math - I wouldn't know how to do that effectively, except for by verifying that the tests make sense and that test coverage is good (which it is).

rgommers commented:
Question: you were careful to define functions with ``nogil`` where possible but never release the GIL anywhere. Are there some computationally heavy parts where you could or intended to do so?

ewmoore commented:
Thanks for the review @rgommers! I'd more or less stopped pushing this along while waiting on #4021.  I have some changes that never made it up to github that clean up checking array sizes and so forth in `qr_insert`.  I'll try to get back to this when I have some time later this week.  

I think it is worth dropping the GIL before entering the computation routines for large enough systems.  I suppose checking the execution time vs. system size in comparison to other places the GIL is released is the way to decide what that threshold should be.

rgommers commented:
gh-4021 seems about ready, so I hope we can merge both of these PRs for the coming release

ewmoore commented:
@rgommers, I believe I've addressed all of your comments.  This will need a rebase once #4021 is merged, and the `?nrm2` call also needs fixed up at that time, but otherwise I think it is complete.

rgommers commented:
Nice, LGTM now.

rgommers commented:
This can be rebased now.

rgommers commented:
I had another look at how this uses the new Cython API, and I think it's indeed preferable to direct linking like was done in ``scipy/cluster/_vq.pyx`` (which needed ``scipy/cluster/cluster_blas.h/pxd``).

ewmoore commented:
Rebased.

rgommers commented:
Okay, 6 days since the message on the mailing list that this is about to be merged. Time to green-button it I think, then it gets to sit in master for a while before branching for the next release.

rgommers commented:
Or better, it could be used inside ``signal.place_poles`` already. @I--P, this is now available.

I--P commented:
@rgommers Nice ! I'm not sure I'll have time to update place-poles with it before 0.16 though.

rgommers commented:
No worries. If you may not be able to do that for a long time, it may be an idea to open an enhancement issue that explains what needs to be done.

argriffing commented:
Should `_decomp_update.pyx` be added to `.gitignore`?  I think the input file is named `_decomp_update.pyx.in`, and `_decomp_update.pyx.c` is already ignored.

rgommers commented:
Indeed. Want to fix that, or should I?

argriffing commented:
You can fix it while I let my scipy build recover from the numpy deprecation flag incompatibility :)

rgommers commented:
done
